src/jit/emitxarch.cpp

```diff
@@ -2901,174 +2901,283 @@ void emitter::emitInsStoreLcl(instruction ins, emitAttr attr, GenTreeLclVarCommo
-    if (dst->isUsedFromMemory())
+    if (dst->isContained() || (dst->isLclField() && (dst->gtRegNum == REG_NA)) || dst->isUsedFromSpillTemp())
     {
+        // dst can only be a reg or modrm
+        assert(dst->isUsedFromMemory() || (dst->gtRegNum == REG_NA) ||
+               instrIs3opImul(ins)); // dst on 3opImul isn't really the dst
         assert(!src->isUsedFromMemory());
-    }
-
-    if (src->isUsedFromMemory())
-    {
-        assert(!dst->isUsedFromMemory());
-    }
-#endif
-
-    // find which operand is a memory op (if any)
-    // and what its base is
-    GenTreeIndir* mem     = nullptr;
-    GenTree*      memBase = nullptr;
 
-    if (dst->isContainedIndir())
-    {
-        mem = dst->AsIndir();
-    }
-    else if (src->isContainedIndir())
-    {
-        mem = src->AsIndir();
-    }
+        memOp = dst;
 
-    if (mem)
-    {
-        memBase = mem->gtOp1;
-    }
-
-    // Find immed (if any) - it cannot be the dst
-    // SSE2 instructions allow only the second operand to be a memory operand.
-    GenTreeIntConCommon* intConst = nullptr;
-    GenTreeDblCon*       dblConst = nullptr;
-    if (src->isContainedIntOrIImmed())
-    {
-        intConst = src->AsIntConCommon();
-    }
-    else if (src->isContainedFltOrDblImmed())
-    {
-        dblConst = src->AsDblCon();
-    }
-
-    // find local field if any
-    GenTreeLclFld* lclField = nullptr;
-    if (src->isLclFldUsedFromMemory())
-    {
-        lclField = src->AsLclFld();
+        if (src->isContained())
+        {
+            assert(src->IsCnsIntOrI());
+            cnsOp = src;
+        }
+        else
+        {
+            otrOp = src;
+        }
     }
-    else if (dst->isLclField() && dst->gtRegNum == REG_NA)
+    else if (src->isContained() || src->isUsedFromSpillTemp())
     {
-        lclField = dst->AsLclFld();
-    }
+        assert(!dst->isUsedFromMemory());
+        otrOp = dst;
 
-    // find contained lcl var if any
-    GenTreeLclVar* lclVar = nullptr;
-    if (src->isLclVarUsedFromMemory())
-    {
-        assert(src->IsRegOptional() || !emitComp->lvaTable[src->gtLclVar.gtLclNum].lvIsRegCandidate());
-        lclVar = src->AsLclVar();
-    }
-    if (dst->isLclVarUsedFromMemory())
-    {
-        assert(dst->IsRegOptional() || !emitComp->lvaTable[dst->gtLclVar.gtLclNum].lvIsRegCandidate());
-        lclVar = dst->AsLclVar();
+        if ((src->IsCnsIntOrI() || src->IsCnsFltOrDbl()) && !src->isUsedFromSpillTemp())
+        {
+            assert(!src->isUsedFromMemory() || src->IsCnsFltOrDbl());
+            cnsOp = src;
+        }
+        else
+        {
+            assert(src->isUsedFromMemory());
+            memOp = src;
+        }
     }
 
-    // find contained spill tmp if any
-    TempDsc* tmpDsc = nullptr;
-    if (src->isUsedFromSpillTemp())
-    {
-        assert(src->IsRegOptional());
-        tmpDsc = codeGen->getSpillTempDsc(src);
-    }
-    else if (dst->isUsedFromSpillTemp())
+    if (memOp != nullptr)
```

Comment:
It might have been better to first handle the simple cases to get those out of the way quicker. As is now you have to scroll through all the `memOp` case to reach the rest.